### PR TITLE
[Feature] Support kenerl updation for some decoder heads.

### PR DESCRIPTION
**Pull request by MengzhangLI**
## Motivation

To support K-Net(https://arxiv.org/abs/2106.14855). In several decoder heads, add `kernel_update` to decide whether (1) feature map before `self.cls_seg` and (2) learnable semantic kernels becoming output for kernel updation. 

## Decoder heads needed to be modified
- [x] ASPPHead
- [x] PSPHead
- [x] FCNHead
- [x] UPerHead

This pr would be merged as soon as possible to proceed the K-Net PR: https://github.com/open-mmlab/mmsegmentation/pull/1289.

### CODE REVIEW

File: mmseg/models/decode_heads/aspp_head.py [hunk 1]
```diff
@@ -103,6 +104,14 @@ def forward(self, inputs):
...
+        feats = self.bottleneck(aspp_outs)
+        output = self.cls_seg(feats)
+        seg_kernels = self.conv_seg.weight.clone()
+        seg_kernels = seg_kernels[None].expand(
+            feats.size(0), *seg_kernels.size())
+        return output, feats, seg_kernels
+
+    def forward(self, inputs):
+        """Forward function."""
+        output, feats, seg_kernels = self.forward_feature(inputs)
```
Comment (MengzhangLI) [inline mmseg/models/decode_heads/aspp_head.py:116]:
```suggestion
        output = self.forward_feature(inputs)
```

File: mmseg/models/decode_heads/aspp_head.py [hunk 2]
```diff
@@ -91,7 +91,7 @@ def __init__(self, dilations=(1, 6, 12, 18), **kwargs):
             norm_cfg=self.norm_cfg,
             act_cfg=self.act_cfg)
 
-    def forward(self, inputs):
+    def forward_feature(self, inputs):
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/aspp_head.py:94]:
```suggestion
    def _forward_feature(self, inputs):
```

File: mmseg/models/decode_heads/fcn_head.py [hunk 3]
```diff
@@ -72,11 +72,16 @@ def __init__(self,
                 norm_cfg=self.norm_cfg,
                 act_cfg=self.act_cfg)
 
-    def forward(self, inputs):
+    def forward_feature(self, inputs):
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/fcn_head.py:75]:
```suggestion
    def _forward_feature(self, inputs):
```

File: mmseg/models/decode_heads/aspp_head.py [hunk 4]
```diff
@@ -91,8 +91,17 @@ def __init__(self, dilations=(1, 6, 12, 18), **kwargs):
             norm_cfg=self.norm_cfg,
             act_cfg=self.act_cfg)
 
-    def forward(self, inputs):
-        """Forward function."""
+    def _forward_feature(self, inputs):
+        """Forward function for feature maps before classifying each pixel with
+        `self.cls_seg` fc.
```
Comment (MeowZheng) [inline mmseg/models/decode_heads/aspp_head.py:96]: in docstring we use `` not `

**Review by MeowZheng — APPROVED** [review]
minor revise

File: mmseg/models/decode_heads/uper_head.py [hunk 5]
```diff
@@ -101,7 +109,7 @@ def forward(self, inputs):
         used_backbone_levels = len(laterals)
         for i in range(used_backbone_levels - 1, 0, -1):
             prev_shape = laterals[i - 1].shape[2:]
-            laterals[i - 1] = laterals[i - 1] + resize(
+            laterals[i - 1] += resize(
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/uper_head.py:112]: Keep the op with before to avoid inplace op, i.e. `+=`.

**Review by Junjun2016 — COMMENTED** [review]
How about other decode_heads?

**Comment by MengzhangLI** [comment]
> How about other decode_heads?

Keep others current contents while these four decoder heads would add docstrings in `self._forward_feature()` to remind it would be used for downstream usage.

**Pull request merged**